File: Heinloo/Chunks/ms2to3/lib/chunktype.py
import struct
import collections

from lib.varint import encode_varint, decode_varint

_chunk_registry = {}

class InvalidChunk(Exception):
    pass

class UnsupportedChunk(Exception):
    pass

class _Chunk(object):
    def _encode(self):
        if self.__hasdata:
            data = struct.pack(self.__layout, *self[:-1]) + self.data

        else:
            data = struct.pack(self.__layout, *self)

        return encode_varint(self.__key) + encode_varint(len(data)) + data

    @classmethod
    def _decode(cls, data):
        if cls.__hasdata:
            kwargs = {'data': data[cls.__parsed_length:]}

        else:
            kwargs = {}

        return cls(*struct.unpack(cls.__layout, data[:cls.__parsed_length]), **kwargs)

def ChunkType(name, key, layout, *fields):
    hasdata = (len(fields) > 0 and fields[-1] == 'data')
    cls = type(name, (_Chunk, collections.namedtuple(name, fields)),
                     {'_Chunk__key': key,
                      '_Chunk__layout': layout,
                      '_Chunk__parsed_length': struct.calcsize(layout),
                      '_Chunk__hasdata': hasdata})

    _chunk_registry[key] = cls
    return cls

def decode_chunk(stream):
    key = decode_varint(stream)
    data_len = decode_varint(stream)

    if data_len > 1024*1024:
        raise InvalidChunk("key=%d, length=%d" % (key, data_len))

    data = stream.read(data_len)

    try:
        cls = _chunk_registry[key]

    except KeyError:
        raise UnsupportedChunk("key=%d, length=%d" % (key, data_len))

    return cls._decode(data)

File: Heinloo/Chunks/javascript/src/chunktype.js
import * as varint from 'varint'
import * as struct from 'python-struct';

var decoder = {}

export function InvalidChunk(message) {
	this.message = message
}

InvalidChunk.prototype = new Error

export function UnsupportedChunk(message) {
	this.message = message
}

UnsupportedChunk.prototype = new Error

export function ChunkType(name, key, layout) {
	var fields = Array.prototype.slice.call(arguments, 3)
	var parsedLength = struct.sizeOf(layout)
	var hasdata = false

	if (fields.length > 0 && fields[fields.length-1] == 'data') {
		fields.pop()
		hasdata = true
	}

	var proto = {_name: name, _length: parsedLength+2, _type: _type, _encode: _encode, toString: toString}

	function _type(attr) {
		var chunk = Object.assign(Object.create(proto), attr)

		if (hasdata) {
			chunk._length += chunk.data.byteLength
		}

		for (let i=chunk._length; i>127; i>>=7)
			chunk._length += 1

		for (let i=key; i>127; i>>=7)
			chunk._length += 1

		return chunk
	}

	function _decode(data, offset) {
		var chunk = Object.create(proto)
		var values = struct.unpack(layout, data.slice(offset, offset+parsedLength))

		for (var i in fields) {
			chunk[fields[i]] = values[i]
		}

		if (hasdata) {
			chunk.data = data.slice(offset+parsedLength)
			chunk._length += chunk.data.byteLength
		}

		for (let i=chunk._length; i>127; i>>=7)
			chunk._length += 1

		for (let i=key; i>127; i>>=7)
			chunk._length += 1

		return chunk
	}

	function _encode() {
		var values = []

		for (var i in fields) {
			values.push(this[fields[i]])
		}

		var data = struct.pack(layout, values)

		if (hasdata) {
			var ekey = varint.encode(key)
			var elen = varint.encode(data.length + this.data.length)
			var result = new Uint8Array(ekey.length + elen.length + data.length + this.data.length)
			result.set(ekey)
			result.set(elen, ekey.length)
			result.set(data, ekey.length + elen.length)
			result.set(this.data, ekey.length + elen.length + data.length)
		}
		else {
			var ekey = varint.encode(key)
			var elen = varint.encode(data.length)
			var result = new Uint8Array(ekey.length + elen.length + data.byteLength)
			result.set(ekey)
			result.set(elen, ekey.length)
			result.set(data, ekey.length + elen.length)
		}

		return result
	}

	function toString() {
		return name + "(" + JSON.stringify(this) + ")"
	}

	decoder[key] = _decode
	return _type
}

export function decode(data, offset) {
	var key = varint.decode(data, offset)
	var bytes = varint.decode.bytes
	var length = varint.decode(data, offset + bytes)
	bytes += varint.decode.bytes

	if (length > 1024*1024) {
		throw new InvalidChunk("key=" + key + ", length=", length)
	}

	decode.bytes = bytes + length

	if (!(key in decoder)) {
		throw new UnsupportedChunk("key=" + key + ", length=", length)
	}

	return decoder[key](data, offset + bytes)
}

